### PR TITLE
Correctly handle rename index on older mysql servers

### PR DESCRIPTION
**Pull request by sgiehl**
### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by bx80 — DISMISSED** [review]
I run MariaDB so I was able to test the `DoesDBSupportRenameIndex()` method, which worked well. 

At some point it would be nice to abstract some of the db capabilities to the Db object so we can do something like `Db::canRenameIndex() === true`, we have tickets for this already.

Since we don't know what other MySQL clones people may be using (Percona?) Would it perhaps be safer to specifically check for MariaDB >= 10.5.2, MySQL >= 5.7 and then return false for anything else?

```
if (strpos(strtolower($databaseVersion), 'mariadb')) {
    return version_compare($databaseVersion, '10.5.2', '>=');
}
if (strpos(strtolower($databaseVersion), 'mysql')) {
    return version_compare($databaseVersion, '5.7', '>=');
}
return false; // something other db
```

**Comment by sgiehl** [comment]
@bx80 I actually had implemented something like that before, but thought it would be too much of an edge case.
Anyway. I had some more thoughts and will rewrite the full code again to not rely on any version detection.

**Comment by sgiehl** [comment]
I have now rewritten this PR. Instead of checking for any server versions it will now always try to rename the index if the old one is correct. Possible syntax errors on old mysql versions will be ignored.

Queries to create the new index and to drop the old index will be always performed, ignoring errors for already or not existing keys.

So in case an correct old key exists: 

* For older SQL servers, it will silently fail to rename the key and the queries to create the new and drop the old index will succeed
* For newer SQL servers, it will rename the key and the queries to create the new and drop the old index will silently fail (as the new key already exists and the old key doesn't)

If there isn't an old key it will only create the new one and drop the old one.

File: core/Updates/5.0.0-b1.php [hunk 1]
```diff
@@ -89,18 +89,23 @@ private function appendLogVisitTableMigrations($migrations)
...
+
+        if ($this->hasCorrectlySetOldIndex()) {
+            // already existing index has the correct fields. Try renaming, but ignore syntax error thrown if rename command does not exist
             $migrations[] = $this->migration->db->sql(
-                "ALTER TABLE `{$this->tableName}` ADD INDEX `{$this->newIndexName}` (`idsite`, `idvisitor`, `visit_last_action_time` DESC)",
-                [DbAlias::ERROR_CODE_DUPLICATE_KEY, DbAlias::ERROR_CODE_KEY_COLUMN_NOT_EXISTS]
-            );
-            $migrations[] = $this->migration->db->dropIndex('log_visit', $this->indexName);
+                "ALTER TABLE `{$this->tableName}` RENAME INDEX `{$this->indexName}` TO `{$this->newIndexName}`",
+                [DbAlias::ERROR_CODE_SYNTAX_ERROR]);
```
Comment (sgiehl) [inline core/Updates/5.0.0-b1.php:99]: Note: I've tested that on an older MySQL and MariaDB. Both return a syntax error if they don't support RENAME INDEX.

**Review by michalkleiner — APPROVED** [review]
The approach looks ok to me even though a bit unconventional to rely on silenced syntax errors, but if it does the job done then why not.

**Pull request merged**